### PR TITLE
[SPARK-40849][SS] Async log purge

### PR DESCRIPTION
**Pull request by jerrypeng**
### What changes were proposed in this pull request?

Purging old entries in both the offset log and commit log will be done asynchronously.

 

For every micro-batch, older entries in both offset log and commit log are deleted. This is done so that the offset log and commit log do not continually grow.  Please reference logic here

 

https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala#L539 

 

The time spent performing these log purges is grouped with the “walCommit” execution time in the StreamingProgressListener metrics.  Around two thirds of the “walCommit” execution time is performing these purge operations thus making these operations asynchronous will also reduce latency.  Also, we do not necessarily need to perform the purges every micro-batch.  When these purges are executed asynchronously, they do not need to block micro-batch execution and we don’t need to start another purge until the current one is finished.  The purges can happen essentially in the background.  We will just have to synchronize the purges with the offset WAL commits and completion commits so that we don’t have concurrent modifications of the offset log and commit log.

### Why are the changes needed?

Decrease microbatch processing latency

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Unit tests

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
cc. @zsxwing @xuanyuanking @viirya Appreciate your reviews. Thanks!

**Comment by HeartSaVioR** [comment]
I'll find a time to review in tomorrow.

**Review by HeartSaVioR — APPROVED** [review]
+1
Let me wait for a couple more days to seek more eyes of review. I'll merge this in early next week if there is no outstanding comment.

**Comment by jerrypeng** [comment]
thanks @HeartSaVioR !

**Comment by zhengruifeng** [comment]
the failed python code gen check is unrelated to this PR, please rebase to make CI green

**Comment by HeartSaVioR** [comment]
https://github.com/jerrypeng/spark/actions/runs/3293047872/jobs/5447874146

Remaining steps are unrelated to this PR - only license check which is respected in this PR.

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.

**Pull request closed without merging**
